[Lin Xingchen] iP

Linxcathyyy opened the pull request:
# Meow
As David Allen said ([source](https://dansilvestre.com/productivity-quotes/)):
> "Your mind is for having ideas, not holding them."

Meow frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/Linxcathyyy/ip/).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :smiling_face_with_three_hearts:

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
#
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Linxcathyyy committed:
feat(Great, Echo, Exit): added great, echo, and exit features for the chat bot

Linxcathyyy committed:
feat(Add, List): added add and list features of the chat bot

Linxcathyyy committed:
feat(Mark as Done): Add Mark as Done feature for chat bot

Linxcathyyy committed:
fix(Mark as Done): Add javadoc for Level-3

Linxcathyyy committed:
feat(Todo, Event, Deadline): Add todo, event and deadline features of the chat bot

Linxcathyyy committed:
feat(A-TextUiTesting): Test all the features of the chat bot

Linxcathyyy committed:
feat(Handle Errors): Add error handling methods for unexpected input

Linxcathyyy committed:
feat(Handle Errors): Add error handling methods for invalid input

Linxcathyyy committed:
feat(Delete): Add delete feature for the chat bot

Linxcathyyy committed:
Add 'save' feature to this chat bot

Linxcathyyy committed:
Update gitignore

Linxcathyyy committed:
Add 'Dates and Times' feature to this chat bot

Linxcathyyy committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Meow.java
#	src/main/java/MeowException.java

Linxcathyyy committed:
Add more OOP to the chat bot program

Linxcathyyy committed:
Add package

Linxcathyyy committed:
Add JUnit tests

Linxcathyyy committed:
Add a JAR file

Linxcathyyy committed:
Refine coding style

Linxcathyyy committed:
Add JavaDoc comments

Linxcathyyy committed:
Add 'Find' feature to the chat bot

Linxcathyyy committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/meow/Parser.java

Linxcathyyy committed:
Merge branch 'branch-Level-9'

Linxcathyyy committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Add Gradle support

Linxcathyyy committed:
Set up gradle for this chat bot

Linxcathyyy committed:
Use checkStyle to detect coding style violations

Linxcathyyy committed:
Add GUI to this java application

Linxcathyyy committed:
Add varargs to find function to providing a multiple keywords search solution

yongxiangng commented on src/main/java/meow/Meow.java:
```diff
@@ -0,0 +1,151 @@
+            default: {
+                throw new InvalidInputException();
+            }
+            }
```
Should the switch statements be wrapped in braces?
https://se-education.org/guides/conventions/java/intermediate.html#layout

yongxiangng commented on src/main/java/meow/Parser.java:
```diff
@@ -0,0 +1,147 @@
+        String validator = "[0-9]{1,2}(/|-)[0-9]{1,2}(/|-)[0-9]{4}";
+        boolean dateValidated = inputDate.matches(validator);
+        int time = Integer.parseInt(inputTime);
+        boolean timeValidated = time >= 0 && time <= 2359;
```
Perhaps using `isValidDate` and `isValidTime` would be a more intuitive name for the boolean variables

yongxiangng commented on src/main/java/meow/Ui.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Prints the greeting message from the chat bot cat Meow.
+     */
+    public String greet() {
+        return "Meow: Hi human, I'm your cat Meow~ What can I do for you?";
+    }
+
+    /**
+     * Prints the goodbye message from the chat bot cat Meow.
+     */
+    public String exit() {
+        return "Meow: Bye human, see you soon! Your cat Meow is going to sleep now~";
+    }
```
You may want to document on the value returned.
```java
/**
 * Some comment...
 *
 * @return Comment on the return value...
 */
public String greet() {
    ...
}
```
https://se-education.org/guides/conventions/java/intermediate.html#comments

yongxiangng commented on src/main/java/meow/Storage.java:
```diff
@@ -0,0 +1,159 @@
+            if (completeStatus.equals("1")) {
+                newTodo.markAsDone();
+            }
+            this.tasksList.add(newTodo);
+            break;
+        }
+        case "D": {
+            String dateAndTime = values[3].trim();
+            String[] dateTimeArray = dateAndTime.split(" ");
+            String date = "";
+            for (int i = 0; i < dateTimeArray.length - 1; i++) {
+                if (i == 1 && Integer.parseInt(dateTimeArray[i]) < 10) {
+                    date = date + "0" + dateTimeArray[i] + " ";
+                } else if (i == 2) {
+                    date = date + dateTimeArray[i];
+                } else {
+                    date = date + dateTimeArray[i] + " ";
+                }
+            }
+            LocalDate localDate = LocalDate.parse(date, DateTimeFormatter.ofPattern("MMM d yyyy"));
+            Deadline newDeadline = new Deadline(task, localDate, dateTimeArray[3]);
+            if (completeStatus.equals("1")) {
+                newDeadline.markAsDone();
+            }
+            this.tasksList.add(newDeadline);
+
+            break;
+        }
+        case "E": {
+            String date = values[3].trim();
+            Event newEvent = new Event(task, date);
+            if (completeStatus.equals("1")) {
+                newEvent.markAsDone();
+            }
+            this.tasksList.add(newEvent);
+            break;
+        }
+        default: {
+            break;
+        }
```
Similar suggestion as the previous switch statement

yongxiangng reviewed (commented):
I like how you followed the style guide for the most parts and your code was well documented.
Although not part of the coding conventions provided, some files are missing new lines at the end and you might want to conveniently add them in while making your edits. 😺

Linxcathyyy committed:
Add assertion to check for null value

Linxcathyyy committed:
Improve overall code quality

The load method in the Storage class contains magic number, and it is
too long.

This way of writing the code makes the load method less readable.

Declaring a constant to represent the magic number and using the SLAP
principle improve code quality and readability.

Let's,
* declare a constant value
* simplify the load method and extract out the convertToLocalDate
method.

Linxcathyyy committed:
Merge branch 'branch-A-CodeQuality'

Linxcathyyy committed:
Add java stream to find function

Linxcathyyy committed:
Merge branch 'branch-A-Streams'

Linxcathyyy committed:
Add CI

Linxcathyyy committed:
Merge branch 'branch-A-CI'

Linxcathyyy committed:
Add C-BetterSearch

Linxcathyyy committed:
Merge pull request #1 from Linxcathyyy/branch-A-Extension

Add C-BetterSearch

Linxcathyyy committed:
Merge branch 'master' of ssh://github.com/Linxcathyyy/ip

Linxcathyyy committed:
Check for stylecheck

Linxcathyyy committed:
improve GUI

Linxcathyyy committed:
improve code style

Linxcathyyy committed:
fix testcase

Linxcathyyy committed:
update gradle.yml for CI

Linxcathyyy committed:
Set theme jekyll-theme-architect

Linxcathyyy committed:
Add user guide

Linxcathyyy committed:
Merge branch 'master' of ssh://github.com/Linxcathyyy/ip

Linxcathyyy committed:
Create release
